### PR TITLE
Fix support for Str datatype Enum signals

### PR DESCRIPTION
**Pull request by DiamondJoseph**
Allows for signals that are backed by enums but have a Signal datatype of str.

e.g. for the Aravis TriggerMode

### CODE REVIEW

**Review by Tom-Willemsen — COMMENTED** [review]
The code, as far as I understand it, looks fine.

Can we add some documentation about exactly how enums are supported and which combinations of `SomeEnum(str, Enum)`, `SomeEnum(Enum)` etc are supported?

e.g. [some other places](https://github.com/DiamondLightSource/dodal/pull/467#discussion_r1582881725) seem to believe that str enums are always required. If that's no longer true can we say so in the docs somewhere.

The testing for this is also getting a little bit hard to follow and check that it's really checking all the cases we care about. Maybe out of scope for this issue though.

**Comment by DiamondJoseph** [comment]
> Can we add some documentation about exactly how enums are supported and which combinations of `SomeEnum(str, Enum)`, `SomeEnum(Enum)` etc are supported?
> 
> e.g. [some other places](https://github.com/DiamondLightSource/dodal/pull/467#discussion_r1582881725) seem to believe that str enums are always required. If that's no longer true can we say so in the docs somewhere.

`SomeEnum(str, Enum)`, and `str` are valid datatypes for a signal with an enumeration as a backend. `SomeEnum(int, Enum)`, `SomeEnum(float, Enum)`, `SomeEnum(Enum)`, `float`, `int` etc. are not.

> The testing for this is also getting a little bit hard to follow and check that it's really checking all the cases we care about. Maybe out of scope for this issue though.

I think https://github.com/bluesky/ophyd-async/pull/137 simplifies the tests by making all of the expected descriptors much simpler to generate, I started following it up again today.

**Comment by Tom-Willemsen** [comment]
> SomeEnum(str, Enum), and str are valid datatypes for a signal with an enumeration as a backend. SomeEnum(int, Enum), SomeEnum(float, Enum), SomeEnum(Enum), float, int etc. are not.

Perfect. So [where the docs currently talk about using `Enum`](https://blueskyproject.io/ophyd-async/main/how-to/make-a-simple-device.html#readable) can you add this sort of clarification? Happy for it to be a link to a new/separate docs page if that's easiest. The relationship between datatype and backend probably also could do with explaining for an unfamiliar user.

File: src/ophyd_async/epics/_backend/_aioca.py [hunk 1]
```diff
@@ -88,11 +88,15 @@ def write_value(self, value: Union[Enum, str]):
             return value
 
     def value(self, value: AugmentedValue):
-        return self.enum_class(value)
+        return value
```
Comment (coretl) [inline src/ophyd_async/epics/_backend/_aioca.py:91]: This means that:
```python
class MyEnum(str, Enum):
   aaa = "AAA"
   bbb = "BBB"

sig = epics_signal_rw(MyEnum, pv)
await sig.connect()
assert (await sig.get_value()) == MyEnum.aaa
assert (await sig.get_value()) == "AAA"
```
Unfortunately
```python
print(await sig.get_value())  # prints "AAA" not <MyEnum.aaa: "AAA">
```
@DominicOram is this ok?

**Review by coretl — COMMENTED** [review]
I'm not convinced that we want to start returning `str` from `SignalRW[Enum].get_value()`, which I think this does. Can we leave the enum conversion in, but make it optional? Or make a new converter for it?

File: src/ophyd_async/epics/_backend/_aioca.py (hunk 1 above)
Comment (DiamondJoseph) [reply src/ophyd_async/epics/_backend/_aioca.py]: Added some tests to ensure that:
for signals backed by an enum:

Signal[str].get_value() is not Enum.val
Signal[Enum].get_value() is Enum.val

in each case, Signal.get_value() == "val" and Signal.get_value() == Enum.val

File: src/ophyd_async/epics/_backend/common.py [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
+            raise TypeError(
+                (
+                    f"{pv} has choices {pv_choices}, "
+                    f"which do not match {datatype}, which has {choices}"
+                )
             )
-        )
-    return datatype
+        return {x: datatype(x) for x in pv_choices}
+    return {x: x for x in pv_choices}
```
Comment (coretl) [inline src/ophyd_async/epics/_backend/common.py:25]: I literally have no idea how this is working, here you drop the reference to the Enum, and keep only its string value...

File: tests/epics/test_signals.py [hunk 3]
```diff
@@ -535,3 +536,26 @@ def test_signal_helpers():
...
+
+
+async def test_str_enum_returns_enum(ioc: IOC):
+    await ioc.make_backend(MyEnum, "enum")
+    pv_name = f"{ioc.protocol}://{PV_PREFIX}:{ioc.protocol}:enum"
+
+    sig = epics_signal_rw(MyEnum, pv_name)
+    await sig.connect()
+    val = await sig.get_value()
+    assert val is MyEnum.b
```
Comment (coretl) [inline tests/epics/test_signals.py:548]: ... but here the value `is` the Enum. How does that work?

File: tests/epics/test_signals.py [hunk 4]
```diff
@@ -535,3 +536,26 @@ def test_signal_helpers():
...
     assert execute._backend.write_pv == "Execute"
+
+
+async def test_str_enum_returns_enum(ioc: IOC):
+    await ioc.make_backend(MyEnum, "enum")
+    pv_name = f"{ioc.protocol}://{PV_PREFIX}:{ioc.protocol}:enum"
+
+    sig = epics_signal_rw(MyEnum, pv_name)
+    await sig.connect()
+    val = await sig.get_value()
```
Comment (coretl) [inline tests/epics/test_signals.py:547]: For my sanity please could we also add this test:
```suggestion
    val = await sig.get_value()
    assert repr(val) == "<MyEnum.b: 'Bbb'>"
```

File: src/ophyd_async/epics/_backend/common.py (hunk 2 above)
Comment (DiamondJoseph) [reply src/ophyd_async/epics/_backend/common.py:25]:
```
if issubclass(datatype, Enum):
        choices = tuple(v.value for v in datatype)
        if choices != pv_choices:
            raise TypeError(
                (
                    f"{pv} has choices {pv_choices}, "
                    f"which do not match {datatype}, which has {choices}"
                )
            )
        return {x: datatype(x) for x in pv_choices}  # <- This is if the datatype is enum, we make a map of str : Enum
    return {x: x for x in pv_choices}  # <- This is if the datatype is not None, not an Enum, is str or some other str extending type. We make a map of str: str
```

File: tests/epics/test_signals.py (hunk 3 above)
Comment (DiamondJoseph) [reply tests/epics/test_signals.py:548]: This is where datatype is an enum, so get_supported_values was map{str:enum} and the EnumConverter has the Enum values keyed by name (or index, appropriately)

**Review by coretl — APPROVED** [review]
Happy to merge when the ordering issue is sorted
